mmeds/util.py

```diff
@@ -1000,55 +1002,87 @@ def make_pheniqs_config(reads_forward, reads_reverse, barcodes_forward, barcodes
+    :mapping_file: File containing sample IDs and barcode pairs
+    :input_dir: Directory containing demultiplexed fastq files
+    :verbose: Print information to stdout
+    """
+    # Read in mapping file
+    output_content = {}
+    map_df = pd.read_csv(Path(mapping_file), sep='\t', header=[0, 1], na_filter=False)
+
+    # Only three columns are needed
+    sample_ids = map_df['#SampleID']['#q2:types']
+    forward_barcodes = map_df['BarcodeSequence']['categorical']
+    reverse_barcodes = map_df['BarcodeSequenceR']['categorical']
 
-                line = f.readline()
+    # Hash data for easy access
+    map_hash = {key: (value1, value2) for (key, value1, value2) in zip(sample_ids, forward_barcodes, reverse_barcodes)}
 
-            # Write output file to output_dir
-            p_out = Path(output_dir) / filename
-            p_out.touch()
-            p_out.write_bytes(gzip.compress(out.encode('utf-8')))
-            print(count, 'written', filename)
+    # All files should follow this naming convention
+    filename_template = '{}_S1_L001_R{}_001.fastq.gz'
+    verbose_template = '{} Reading {}'
+    count = 0
+
+    # Generate output for each sample's forward and reverse input files
+    for key in map_hash:
+        forward_filename = filename_template.format(key, 1)
+        reverse_filename = filename_template.format(key, 2)
+
+        # Forward Reads
+        if verbose:
             count += 1
+            print(verbose_template.format(count, forward_filename))
+        output_content[forward_filename] = get_stripped_file_content(
+            num_allowed_errors,
+            map_hash[key][0],
+            map_hash[key][1],
+            Path(input_dir) / forward_filename
+        )
+
+        # Reverse Reads
+        if verbose:
+            count += 1
+            print(verbose_template.format(count, reverse_filename))
+        output_content[reverse_filename] = get_stripped_file_content(
+            num_allowed_errors,
+            map_hash[key][0],
+            map_hash[key][1],
+            Path(input_dir) / reverse_filename
+        )
+
+    return output_content
+
+
+def get_stripped_file_content(num_allowed_errors, forward_barcode, reverse_barcode, filename):
+    """
+    Get the error-stripped content of one demultiplexed fastq file
+    ==============================================================
+    :num_allowed_errors: Maximum number of errors in barcode pairs to not be stripped
+    :forward_barcode: First of two barcodes associated with the sample
+    :reverse_barcode: Second of two barcodes associated with the sample
+    :filename: Absolute Path object to demultiplexed fastq file
+    """
+    content = ''
+    f = gzip.open(filename, mode='rt')
+
+    # Create a list of every read in the file as tuples, structured as [(whole_read, forward_barcode, reverse_barcode)]
+    reads = re.findall(r'(@M.+:0:([ACTG]+)-([ACTG]+)\n.+\n.+\n.+\n)', f.read())
```

Comment:
Not familiar with how you are processing the file here: you are *not* reading all lines of the file into `reads`, are you? Because then you end up with a gigantic list in memory. Maybe I'm just not understanding the code.

Comment on what that `@M.+:0: ([ACTG...` string means would be helpful. You can write in the comment an example seq ID for others to understand better.

Reply:
can use a regex iterator so it's not all in memory

re.finditer(pattern, string, flags=0)
https://docs.python.org/3/library/re.html